File: desktop_app/test3.py
import cv2
import numpy as np

# Initialize video capture
cap = cv2.VideoCapture('files/Traffic_Laramie_1.mp4')

# Initialize the MOG2 background subtractor with parameters
mog2 = cv2.createBackgroundSubtractorMOG2(history=500, varThreshold=16, detectShadows=False)

# Define morphological operations kernels
kernel = cv2.getStructuringElement(cv2.MORPH_ELLIPSE, (5, 5))
small_kernel = cv2.getStructuringElement(cv2.MORPH_ELLIPSE, (3, 3))

# Set minimum object size threshold
min_object_area = 1000

# Process each frame with MOG2
while True:
    ret, frame = cap.read()
    if not ret:
        break

    # Resize frame for faster processing
    frame = cv2.resize(frame, (640, 360))

    # Apply MOG2 background subtractor to get the foreground mask
    fg_mask = mog2.apply(frame)

    # Apply morphological operations to reduce noise
    fg_mask = cv2.morphologyEx(fg_mask, cv2.MORPH_OPEN, kernel)
    fg_mask = cv2.morphologyEx(fg_mask, cv2.MORPH_CLOSE, kernel)

    # Apply erosion to remove small noise regions
    fg_mask = cv2.erode(fg_mask, small_kernel, iterations=1)

    # Apply a threshold to remove gray pixels (weak foreground regions)
    _, fg_mask = cv2.threshold(fg_mask, 200, 255, cv2.THRESH_BINARY)

    # Find contours in the foreground mask
    contours, _ = cv2.findContours(fg_mask, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)

    # Filter contours based on minimum object area
    for contour in contours:
        if cv2.contourArea(contour) < min_object_area:
            continue

        # Get bounding box coordinates
        x, y, w, h = cv2.boundingRect(contour)

        # Draw bounding box on the original frame
        cv2.rectangle(frame, (x, y), (x + w, y + h), (0, 255, 0), 2)

    # Display the frame with bounding boxes
    cv2.imshow('Detected Objects', frame)

    if cv2.waitKey(20) & 0xFF == ord('q'):
        break

# Release video capture and destroy all windows
cap.release()
cv2.destroyAllWindows()

File: desktop_app/playground/test3.cpp
#include <opencv2/opencv.hpp>
#include <iostream>

using namespace cv;
using namespace std;

int main() {
    // Initialize video capture
    VideoCapture cap("files/Traffic_Laramie_1.mp4");
    if (!cap.isOpened()) {
        cerr << "Error opening video file" << endl;
        return -1;
    }

    // Initialize the MOG2 background subtractor with parameters
    Ptr<BackgroundSubtractorMOG2> mog2 = createBackgroundSubtractorMOG2(500, 16, false);

    // Define morphological operations kernels
    Mat kernel = getStructuringElement(MORPH_ELLIPSE, Size(5, 5));
    Mat small_kernel = getStructuringElement(MORPH_ELLIPSE, Size(3, 3));

    // Set minimum object size threshold
    int min_object_area = 1000;

    // Process each frame with MOG2
    while (true) {
        Mat frame;
        cap >> frame;
        if (frame.empty()) {
            break;
        }

        // Resize frame for faster processing
        resize(frame, frame, Size(640, 360));

        // Apply MOG2 background subtractor to get the foreground mask
        Mat fg_mask;
        mog2->apply(frame, fg_mask);

        // Apply morphological operations to reduce noise
        morphologyEx(fg_mask, fg_mask, MORPH_OPEN, kernel);
        morphologyEx(fg_mask, fg_mask, MORPH_CLOSE, kernel);

        // Apply erosion to remove small noise regions
        erode(fg_mask, fg_mask, small_kernel, Point(-1, -1), 1);

        // Apply a threshold to remove gray pixels (weak foreground regions)
        threshold(fg_mask, fg_mask, 200, 255, THRESH_BINARY);

        // Find contours in the foreground mask
        vector<vector<Point>> contours;
        findContours(fg_mask, contours, RETR_EXTERNAL, CHAIN_APPROX_SIMPLE);

        // Filter contours based on minimum object area
        for (const auto& contour : contours) {
            if (contourArea(contour) < min_object_area) {
                continue;
            }

            // Get bounding box coordinates
            Rect bounding_box = boundingRect(contour);

            // Draw bounding box on the original frame
            rectangle(frame, bounding_box, Scalar(0, 255, 0), 2);
        }

        // Display the frame with bounding boxes
        imshow("Detected Objects", frame);

        if (waitKey(20) == 'q') {
            break;
        }
    }

    // Release video capture and destroy all windows
    cap.release();
    destroyAllWindows();

    return 0;
}
